src/parsers/primitives.ts

```diff
@@ -1,16 +1,27 @@
 import {getLanguage} from '../languages';
+import type {Language} from '../CodeMirrorBlocks';
 
 export class Primitive {
-  constructor(languageId, name, {argumentTypes, returnType}={}) {
+  languageId: string;
+  name: string;
+  parse: Language['parse'];
+  primitivesFn: Language['primitivesFn'];
+  primitives: Language['primitives'];
+  getASTNodeForPrimitive: Language['getASTNodeForPrimitive'];
+  getLiteralNodeForPrimitive: Language['getLiteralNodeForPrimitive'];
+  argumentTypes: string[];
+  returnType: string;
+
+  constructor(languageId: string, name: string, config: {argumentTypes?: string[], returnType?: string} = {}) {
     this.languageId = languageId;
     this.parse = languageId ? getLanguage(languageId).parse : null;
     this.primitives = languageId ? getLanguage(languageId).primitives : null;
-    this.primitivesFn = languageId ? getLanguage(languageId).PrimitivesFn : null;
+    this.primitivesFn = languageId ? getLanguage(languageId).primitivesFn : null;
     this.getASTNodeForPrimitive = languageId ? getLanguage(languageId).getASTNodeForPrimitive : null;
     this.getLiteralNodeForPrimitive = languageId ? getLanguage(languageId).getLiteralNodeForPrimitive : null;
```

Comment:
This denormalization makes me wonder about the intended semantics. For example, can I _update_ a language definition by calling `addLanguage` more than once, perhaps with a new set of primitives? Will I need to create a brand new CodeMirror instance when I do so?

```js
Languages.addLanguage('my-language', {...});
// ... do some stuff using primitives ...
Languages.addLanguage('my-language', {...});
// ... do some more stuff using primitives
```

Which primitives would I expect to be using after the second call to `Languages.addLanguage`?

Reply:
This predates the languages-as-separate-modules refactoring that we did back in 2020, and assumed that the CMB library itself would be the entry point. In that use-case, someone would need to add their language dynamically in order to register it. 

This is now completely unnecessary, and should probably be removed.